cmd/deviceplugin/macvtap.go

```diff
@@ -7,16 +7,24 @@ import (
 	"github.com/golang/glog"
 	"github.com/kubevirt/device-plugin-manager/pkg/dpm"
 	macvtap "github.com/kubevirt/macvtap-cni/pkg/deviceplugin"
+	"github.com/kubevirt/macvtap-cni/pkg/util"
 )
 
 func main() {
 	flag.Parse()
+	// Device plugin operates with several goroutines that might be
+	// relocated among different OS threads with different namespaces.
+	// We capture the main namespace here and make sure that we do any
+	// network operation on that namespace.
+	// See
+	// https://github.com/containernetworking/plugins/blob/master/pkg/ns/README.md
+	mainNsPath := util.GetMainThreadNetNsPath()
```

Comment:
Thanks for the explanation.

One thing I'm not following is why did this start to manifest with this change. What changed with regard to your first implementation ? 

Is it that now you're exposing more resources, and as such this situation manifests ? E.g. if with the 'old' implementation we had exposed 2 resources (let's say, eth0 & eth1) it would also manifest ?

Reply:
Probably because I added an extra goroutine and that triggered things in a different way in the tests. The problem was already there.

Reply:
We don't need it. It is already handled internally by WithNetNSPath/netns.Do.

Reply:
Explaining better after re-read. `WithNetNSPath` is equivalent to `netns.Do`, it just works with a path to a namespace instead of a namespace object. Both also make appropriate use of `runtime.LockOSThread` internally as needed.
 
`runtime.LockOSThread` is to be used temporarily. We cannot permanently lock threads in a namespace, as we have no control on what are those threads used for and they are long-lived. Temporary use of `runtime.LockOSThread` is precisely what `netns.Do` is handling for us.

Note that also netlink subscription happens inside the needed namespace.